skyportal/models/public_pages/public_source_page.py

```diff
@@ -0,0 +1,160 @@
+        """Get the options check to be displayed on the public page."""
+        photometry = self.data.get("photometry")
+        classifications = self.data.get("classifications")
+        return {
+            "photometry": "no element"
+            if photometry == []
+            else "public"
+            if photometry
+            else "private",
+            "classifications": "no element"
+            if classifications == []
+            else "public"
+            if classifications
+            else "private",
+        }
+
+    def publish(self):
+        """Set the page to public and create the public source page."""
+        self.is_public = True
+        self.generate_public_source_page()
+
+    def generate_public_source_page(self):
+        """Generate a public page of the source and cache it."""
+        data = self.data
+        if isinstance(data, str):
+            try:
+                data = json.loads(data)
+            except Exception:
+                data = {"status": "error", "message": "Invalid JSON data."}
+        if data.get("status") == "error":
+            raise ValueError(data.get("message", "Invalid JSON data."))
+        elif data.get("status") == "pending":
+            raise ValueError("Public source page is still being generated.")
+        # Set the data status to pending to indicate that the page is being generated
+        self.data.update({"status": "pending"})
+
+        cache_key = f"source_{self.source_id}_version_{self.created_at_iso}"
+        public_source_page = self.generate_html(data)
+        cache[cache_key] = dict_to_bytes({"public": True, "html": public_source_page})
+        self.data.update({"status": "success"})
+
+    def generate_html(self, public_data):
+        """Generate HTML for the public page of the source."""
+        if isinstance(public_data, str):
+            public_data = json.loads(public_data)
+
+        # Create the filters mapper
+        if "photometry" in public_data and public_data["photometry"] is not None:
+            from skyportal.handlers.api.photometry import get_bandpasses_to_colors
+
+            filters = {
+                photometry["filter"] for photometry in public_data.get("photometry", [])
+            }
+            public_data["filters_mapper"] = get_bandpasses_to_colors(filters)
+
+        environment = jinja2.Environment(
+            loader=jinja2.FileSystemLoader("./static/public_pages/sources/source")
+        )
+        environment.policies['json.dumps_function'] = to_json
+        template = environment.get_template("source_template.html")
+        html = template.render(
+            data=public_data,
+            creation_date=self.created_at,
+        )
+        return html
+
+    def remove_from_cache(self):
+        """Remove the page from the cache."""
+        cache_key = f"source_{self.source_id}_version_{self.created_at_iso}"
+        del cache[cache_key]
+
+
+@event.listens_for(PublicSourcePage, 'after_insert')
```

Comment:
Is that needed??? We already have a `created_at` column on every data model, I find that very surprising.